### PR TITLE
stubtest: relax async checking for abstract methods

### PR DESCRIPTION
**Pull request by AlexWaygood**
### Description

As has been pointed out in https://github.com/python/typeshed/pull/7475, the current typeshed definitions for various async methods in the stub for `typing` are problematic.

Until recently, these were all "synchronous" methods that returned `Awaitable[X]`. However, they were recently changed to be `async def` methods. This more closely match their definitions at runtime, but is problematic: the classes in `typing.pyi` represent abstract interfaces, and it's not true that an object _has_ to return a coroutine from its `__anext__` method in order to be considered an `AyncIterator` (it just has to return something that's awaitable).

Unfortunately, following https://github.com/python/mypy/pull/12212, stubtest will now emit an error if a method is async at runtime but not in the stub.

I think it makes sense to relax this check for abstract methods, so that's what this PR proposes doing.

## Test Plan

Three new test cases added to `teststubtest.py`.

cc. @hauntsaninja, @graingert

### CODE REVIEW

File: mypy/stubtest.py [hunk 1]
```diff
@@ -734,14 +734,31 @@ def verify_funcitem(
...
+            getattr(runtime, "__isabstractmethod__", False)
+            or (isinstance(stub, nodes.FuncDef) and stub.is_abstract)
+        ):
+            # Be more permissive if the method is an abstractmethod:
+            # Only error if the return type of the stub isn't awaitable
+            if isinstance(stub.type, mypy.types.CallableType):
+                ret_type = mypy.types.get_proper_type(stub.type.ret_type)
+                should_error = (
+                    isinstance(ret_type, mypy.types.Instance)
+                    and "__await__" not in ret_type.type.names
```
Comment (AlexWaygood) [inline mypy/stubtest.py:747]: Is there a more idiomatic way of figuring out whether the return type of a method in the stub is a subtype of `Awaitable[Any]`?

Comment (jhance) [reply mypy/stubtest.py]: I think the unused awaitable stuff does something similar.

Comment (AlexWaygood) [reply mypy/stubtest.py]: Thanks! I've modified my PR to add a helper property in mypy.types to avoid code duplication

**Review by hauntsaninja — COMMENTED** [review]
I'm not sure I fully understand this change, but I'm not yet convinced it's better than an allowlist entry. This seems to encourage LSP violations (nitpicky ones, but still). Note also that abstractmethods in Python can very much have implementations.

**Comment by AlexWaygood** [comment]
> I'm not sure I fully understand this change, but I'm not yet convinced it's better than an allowlist entry.

I see it as similar to the various synchronous abstract methods where we have a looser signature in the stub than the default implementation at runtime.

For example, `collections.abc.Iterator.__iter__` and `contextlib.AbstractContextManager.__enter__` both return `self` at runtime. But we wouldn't want them to be annotated with `-> Self` in the stubs, because these classes are abstract interfaces that represent formal definitions of what it means to be an "iterator" or a "context manager", and a class doesn't _have_ to return `self` from its `__iter__` method order for it to be considered an "iterator".

In the same way, `AsyncIterator.__anext__` is a coroutine function at runtime -- that's the default implementation. But a class's `__anext__` method doesn't _have_ to be a coroutine function in order for it to be considered an "asynchronous iterator" (though it _does_ need to return an awaitable). So it makes sense to have a looser signature for `AsyncIterator.__anext__` than the default implementation, just like it does for `collections.abc.Iterator.__iter__` and `contextlib.AbstractContextManager.__enter__`.

It seems to me that classes with abstract methods are generally going to represent abstract interfaces of some kind. As such, it might often be the case that the default implementation of an abstract method is a coroutine function, even if subclasses are not _necessarily_ expected to implement the method as a coroutine function in order to be considered compatible with the interface. Hence, this PR :)

File: mypy/test/teststubtest.py [hunk 2]
```diff
@@ -224,6 +227,65 @@ def test_coroutines(self) -> Iterator[Case]:
...
+        yield Case(
+            stub="""
+            from abc import abstractmethod
+            from typing import Generator, Any, Coroutine, Awaitable
+            class _CustomAwaitable:
+                def __await__(self) -> Generator[Any, Any, Any]: ...
+            class _CustomAwaitable2(Awaitable[Any]): ...
+            class Foo:
+                @abstractmethod
+                def abstract(self) -> _CustomAwaitable: ...
```
Comment (JelleZijlstra) [inline mypy/test/teststubtest.py:255]: Why do we want to allow this? Doesn't seem particularly useful.

Comment (AlexWaygood) [reply mypy/test/teststubtest.py:255]: I can't remember, and I'm not sure I care anymore. I guess I'll just close this PR :)

**Pull request closed without merging**